### DISCUSSION TITLE
How can I remove X-Powered-By header?

### DISCUSSION THREAD

**Discussion by joelamltc** [directus/directus#16191] [Archive]
### Describe the Bug

Currently, there is no way to turn off the X-Powered-By header.

### To Reproduce

1. lift up directus to up and running
2. inspect network tab in Inspection
<img width="1680" alt="image" src="https://user-images.githubusercontent.com/13143636/198200443-720a38b1-c8a3-4e78-bf1a-4a00ce429cea.png">

**Comment by joelamltc** [comment]
Updates 2022-10-27:
Since there isn't a configuration option for this purpose, currently I am removing the "X-Powered-By" header at the web server level.
e.g. for Apache
```
Header unset "X-Powered-By"
Header unset "x-powered-by"
```

**Comment by azrikahar** [comment]
It should be possible to tap into the express instance via a [custom init hook](https://docs.directus.io/extensions/hooks.html#init) to remove the header (or any other headers if you wish to) 👍 Here's an example:

```js
module.exports = ({ init }) => {
	init('middlewares.after', ({ app }) => {
		app.use((_req, res, next) => {
			res.removeHeader('X-Powered-By');
			next();
		});
	});
};
```

Out of curiosity, can you share the reason to remove the header? Understandably a pretty common concern is security, but AFAIK it's technically more of an issue for cases that shows the specific versions, whereas here it'll only show `Directus` (but I can also see how that would also be perceived as a hint of sort). But more so curious what was motivation in this particular instance 🙂

**Reply by piotr-cz** [reply]
Small quirk is that it's not possible to modify headers for the `ping` endpoint
Might be that it's because it is registered as a middleware (before `middlewares.after` event) and not a route: https://github.com/directus/directus/blob/v9.21.2/api/src/app.ts#L210

**Reply by azrikahar** [reply]
Interesting! I believe it is because `res.send('pong')` (unfortunately) "responded" before the removal of said header can happen because it's placed later. Perhaps the setHeader block can be moved to before `middlewares.before` 🤔

In the meantime, would you mind trying the following workaround and see whether it addresses that scenario for you?

```js
module.exports = ({ init }) => {
	init('app.before', ({ app }) => {
	        /**
	        * this will override the ping endpoint entirely, 
	        * but most importantly it'll be declared before the 
	        * existing `res.setHeader('X-Powered-By', 'Directus')`
	        */
		app.get('/server/ping', (req, res) => {
			res.send('pong');
		});
	});
	init('middlewares.after', ({ app }) => {
		app.use((_req, res, next) => {
			res.removeHeader('X-Powered-By');
			next();
		});
	});
};
```

**Reply by piotr-cz** [reply]
Yes, the above code removes the header.

By the way, I'm not interested in removing the `X-Powered-By` header, but rather in adding custom one.
To do this also for the `ping` route, it's required to use `middlewares.before` event.

**Comment by joelamltc** [comment]
@azrikahar Thanks for the quick response! Yea, it is around the security concern of exposing development platform information to the public, an attacker can take advantage of the known vulnerabilities of the development platform which allows an attacker to exploit existing vulnerabilities and launch further attacks.

**Reply by azrikahar** [reply]
Likewise and thanks for clarifying!

**Comment by Pachat** [comment]
If hiding Directus is the goal, there are other places to look at.
For instance, the "Page source code" shows
```
<body class="auto">
	<noscript>
		<strong>We're sorry but Directus doesn't work without JavaScript enabled. Please enable it to continue.</strong>
	</noscript>

	<div id="app"></div>

	<div id="dialog-outlet"></div>
	<div id="menu-outlet"></div>
</body>
```
from `node_modules/@directus/app/dist/index.html`
_Restart Directus after any modification._
